### PR TITLE
[numpy] `torch.a{cos, tan}` : promote integer inputs to float 

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference #42515

### CODE REVIEW

File: aten/src/ATen/native/UnaryOps.cpp [hunk 1]
```diff
@@ -149,17 +149,24 @@ Tensor& deg2rad_out(Tensor& result, const Tensor& self) {
 Tensor deg2rad(const Tensor& self) { return unary_op_impl(self, at::deg2rad_out); }
 Tensor& deg2rad_(Tensor& self) { return unary_op_impl_(self, at::deg2rad_out); }
 
-Tensor& asin_out(Tensor& result, const Tensor& self) { return unary_op_impl_out(result, self, asin_stub); }
-Tensor asin(const Tensor& self) { return unary_op_impl(self, at::asin_out); }
+Tensor& asin_out(Tensor& result, const Tensor& self) { return unary_op_impl_float_out(result, self, asin_stub); }
+Tensor asin(const Tensor& self) {
+  if (self.is_sparse()) {
```
Comment (mruberry) [inline aten/src/ATen/native/UnaryOps.cpp:154]: Can you elaborate on what's going on here? In particular, how does sparse asin end up here? Maybe it should dispatch to something else?

**Review by mruberry — COMMENTED** [review]
Changes to acos and atan look great and we can merge those separately if you like. I'm not sure what's going on with asin. If sparse asin is dispatched there maybe it needs a dispatch update to be consistent with acos and atan?

File: aten/src/ATen/native/UnaryOps.cpp (hunk 1 above)
Comment (kshitij12345) [reply aten/src/ATen/native/UnaryOps.cpp:154]: Looking at the entry below ->
https://github.com/pytorch/pytorch/blob/ee0033af9b7531ba0a453844dc621f8cfa140c20/aten/src/ATen/native/native_functions.yaml#L634-L650

Only inplace and out variant have a special dispatch for sparse.

So when we call `asin`, we allocate an output tensor and call the out variant.

TensorIterator doesn't work if input tensor is sparse and out is undefined tensor as it throws `RuntimeError: unsupported tensor layout: Sparse` (the path with `unary_op_impl_float`), 
so for sparse we use the previously working path `unary_op_impl` which defines the output tensor to have same dtype and layout as input.

But thinking about that now. It means sparse tensor wouldn't promote from integer to floats. :thinking: 
Will have to look for a better solution.
Thanks for pointing that out!

**Comment by kshitij12345** [comment]
> Changes to acos and atan look great and we can merge those separately if you like. I'm not sure what's going on with asin. If sparse asin is dispatched there maybe it needs a dispatch update to be consistent with acos and atan?

I think it would be better to merge `acos` and `atan` here.  As they don't have a special dispatch on sparse, we can look at asin seperately in a different PR.

https://github.com/pytorch/pytorch/blob/ee0033af9b7531ba0a453844dc621f8cfa140c20/aten/src/ATen/native/native_functions.yaml#L331-L345

https://github.com/pytorch/pytorch/blob/ee0033af9b7531ba0a453844dc621f8cfa140c20/aten/src/ATen/native/native_functions.yaml#L663-L677

**Review by mruberry — APPROVED** [review]
Thanks @kshitij12345!

**Pull request closed without merging**